src/org/intellij/erlang/compilation/ErlangPrepareDependenciesCompileTask.java

```diff
@@ -140,55 +133,83 @@ private static ErlangModuleBuildOrderDescriptor getModuleBuildOrderInner(Module
+      for (Module module : modulesToCompile) {
+        buildForErlangModules(module);
+      }
     }
 
     @NotNull
-    public static ErlangFilesDependencyGraph createSemiGraph(@NotNull Module module,
-                                                             boolean onlyTestModules) {
-      Collection<VirtualFile> erlangModules = ErlangModulesUtil.getErlangModuleFiles(module, onlyTestModules);
-      Collection<VirtualFile> erlangHeaders = ErlangModulesUtil.getErlangHeaderFiles(module, onlyTestModules);
+    public static ErlangFilesDependencyGraph createSemiGraph(@NotNull Module[] modulesToCompile) {
+      return new ErlangFilesDependencyGraph(modulesToCompile);
+    }
+
+    @Override
+    public Collection<String> getNodes() {
+      return myPathsToDependenciesMap.keySet();
+    }
 
-      return new ErlangFilesDependencyGraph(erlangModules, erlangHeaders, getGlobalParseTransform(module), PsiManager.getInstance(module.getProject()));
+    @Override
+    public Iterator<String> getIn(@NotNull String filePath) {
+      return myPathsToDependenciesMap.get(filePath).iterator();
     }
 
-    private void buildForModules(@NotNull Collection<VirtualFile> erlangModules,
-                                 @NotNull List<String> globalParseTransforms) {
+    @NotNull
+    public List<String> getDependencies(@NotNull String filePath) {
+      List<String> dependencies = myPathsToDependenciesMap.get(filePath);
+      assert dependencies != null;
+      return dependencies;
+    }
+
+    @NotNull
+    private static List<VirtualFile> collectHeaders(@NotNull Module[] modulesToCompile) {
+      List<VirtualFile> erlangHeaders = ContainerUtil.newArrayList();
+      for (Module module : modulesToCompile) {
+        erlangHeaders.addAll(ErlangModulesUtil.getErlangHeaderFiles(module, false));
+        erlangHeaders.addAll(ErlangModulesUtil.getErlangHeaderFiles(module, true));
+      }
+      return erlangHeaders;
+    }
+
+    private void buildForErlangModules(@NotNull Module module) {
+      List<String> globalParseTransform = getGlobalParseTransform(module);
+      buildForErlangModules(ErlangModulesUtil.getErlangModuleFiles(module, false), globalParseTransform);
+      buildForErlangModules(ErlangModulesUtil.getErlangModuleFiles(module, true), globalParseTransform);
+    }
+
+    private void buildForErlangModules(@NotNull Collection<VirtualFile> erlangModules,
+                                       @NotNull List<String> globalParseTransforms) {
       for (VirtualFile erlangModule : erlangModules) {
         Set<String> dependencies = ContainerUtil.newHashSet();
         ErlangFile erlangFile = getErlangFile(erlangModule);
         addDeclaredDependencies(erlangFile, dependencies);
-        dependencies.addAll(getPathsFromNames(globalParseTransforms));
+        dependencies.addAll(resolvePathsFromNames(globalParseTransforms));
         myPathsToDependenciesMap.put(erlangModule.getPath(), ContainerUtil.newArrayList(dependencies));
       }
     }
```

Comment:
```buildForHeaders``` is almost exactly the same as this method. I believe, there should be a single method responsible for collecting dependency information.

It also goes for other parts where you split building your graph for modules and headers. You do nearly the same stuff, but in different ways. For example, you write 
```java
      for (Module module : modulesToCompile) {
        buildForErlangModules(module);
      }
```
for modules, but 
```java
      List<VirtualFile> erlangHeaders = collectHeaders(modulesToCompile);
      buildForHeaders(erlangHeaders);
```
for headers. Overall, it makes reading your code more complicated. Please, try to do similar things with similar code, when you can't avoid duplication.